### PR TITLE
Stateless real-time GET

### PR DESCRIPTION
**Pull request by pxsalehi**
For real-time get on Stateless, we'd need to first check the indexing shard whether it has 
the document in its Translog, if not we might have to wait on the search shard and then 
handle the GET locally. 

Relates ES-5537

### CODE REVIEW

**Comment by pxsalehi** [comment]
Thanks for the feedback @henningandersen.

> I think there is no way around contacting the indexing shard for real-time GET.

So as long as the GET request has its `realtime` field set to true (which is the default) we'd forward it to the indexing shard?

**Comment by pxsalehi** [comment]
> So as long as the GET request has its realtime field set to true (which is the default) we'd forward it to the indexing shard?

We discussed this on another channel with Henning. The disadvantage here would be that this might cause a lot of load on the indexing shards. We'd rather only contact the indexing shards for a Translog lookup.

File: server/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java [hunk 1]
```diff
@@ -175,24 +192,31 @@ public void start() {
...
             if (lastFailure == null || TransportActions.isReadOverrideException(currentFailure)) {
                 lastFailure = currentFailure;
                 this.lastFailure = currentFailure;
             }
-            final ShardRouting shardRouting = shardIt.nextOrNull();
+            this.lastResponse = currentResponse;
+
+            ShardRouting shardRouting = shardIt.nextOrNull();
+            if (shardRouting != null) {
+                while (shardRouting != null && skipShardCopy(shardRouting, lastResponse, lastFailure)) {
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java:216]: NIT: it seems this `if` is redundant

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java:216]: I guess it is! But I think most of this is going to change anyway!

**Comment by pxsalehi** [comment]
I'm going to rework the PR, and scope it only to GET for now.

File: server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java [hunk 2]
```diff
@@ -85,18 +83,14 @@ protected boolean resolveIndex(MultiGetShardRequest request) {
 
     @Override
     protected ShardIterator shards(ClusterState state, InternalRequest request) {
-        ShardIterator shards = clusterService.operationRouting()
+        ShardIterator iterator = clusterService.operationRouting()
             .getShards(state, request.request().index(), request.request().shardId(), request.request().preference());
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java:87]: Could it be null here as well?

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 3]
```diff
@@ -71,24 +69,18 @@ protected boolean resolveIndex(GetRequest request) {
...
-        if (DiscoveryNode.isStateless(clusterService.getSettings())) {
-            return new PlainShardIterator(
-                shards.shardId(),
-                shards.getShardRoutings().stream().filter(ShardRouting::isPromotableToPrimary).collect(Collectors.toList())
-            );
-        } else {
-            return shards;
+        if (iterator == null) {
+            return null;
         }
```
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:85]: Yeah it is handled in `org.elasticsearch.action.support.single.shard.TransportSingleShardAction.AsyncSingleAction#start`.

File: server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java (hunk 2 above)
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java:87]: Yes. Good point. This is how our current workaround works. There should be a null check there. I'll adjust it with this PR.

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java:87]: I opened https://github.com/elastic/elasticsearch/pull/94164 for this.

File: rest-api-spec/src/yamlRestTest/java/org/elasticsearch/test/rest/ClientYamlTestSuiteIT.java [hunk 4]
```diff
@@ -30,6 +30,7 @@ public class ClientYamlTestSuiteIT extends ESClientYamlSuiteTestCase {
         .module("mapper-extras")
         .feature(FeatureFlag.TIME_SERIES_MODE)
         .feature(FeatureFlag.DLM_ENABLED)
+        .nodes(2)
```
Comment (pxsalehi) [inline rest-api-spec/src/yamlRestTest/java/org/elasticsearch/test/rest/ClientYamlTestSuiteIT.java:33]: I have two open PRs for the YAML cluster size and the YAML tests. Once merged, it will reduce the scope of this PR.

**Comment by pxsalehi** [comment]
@elasticmachine update branch

**Comment by pxsalehi** [comment]
@elasticmachine update branch

**Comment by pxsalehi** [comment]
This is ready for review.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 5]
```diff
@@ -104,16 +105,72 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
...
+            }
+        }
+    }
+
+    private void handleGetOnUnpromotableShard(GetRequest request, IndexShard indexShard, ActionListener<GetResponse> listener)
+        throws IOException {
+        ShardId shardId = indexShard.shardId();
+        var node = clusterService.state()
+            .nodes()
+            .get(clusterService.state().routingTable().shardRoutingTable(shardId).primaryShard().currentNodeId());
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:137]: `indexShard.getReplicationGroup().getRoutingTable().primaryShard().currentNodeId()`

Does that work?

**Review by Tim-Brooks — COMMENTED** [review]
This mostly looks fine. But I would like to have one more brief chat about this tomorrow to make sure I understand a few specific things.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java (hunk 5 above)
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java]: I think you can only do that if you are a primary. `handleGetOnUnpromotableShard` runs always on a non-primary.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 6]
```diff
@@ -104,16 +105,72 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
     protected void asyncShardOperation(GetRequest request, ShardId shardId, ActionListener<GetResponse> listener) throws IOException {
         IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
         IndexShard indexShard = indexService.getShard(shardId.id());
-        if (request.realtime()) { // we are not tied to a refresh cycle here anyway
-            super.asyncShardOperation(request, shardId, listener);
+        var shardRouting = clusterService.state().getRoutingNodes().node(clusterService.localNode().getId()).getByShardId(shardId);
```
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java]: I've pushed https://github.com/elastic/elasticsearch/pull/93976/commits/f795692f637821e0b41bed78e360cd22312f408b and also handle a few corner-cases there.

File: server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java [hunk 7]
```diff
@@ -90,16 +88,12 @@ protected ShardIterator shards(ClusterState state, InternalRequest request) {
...
-                iterator.shardId(),
-                iterator.getShardRoutings().stream().filter(ShardRouting::isPromotableToPrimary).collect(Collectors.toList())
-            );
-        } else {
-            return iterator;
-        }
+        return new PlainShardIterator(
+            iterator.shardId(),
+            iterator.getShardRoutings().stream().filter(ShardRouting::isPromotableToPrimary).toList()
+        );
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java:96]: This change looks unrelated and also looks like something we will touch anyway in the mget? Perhaps we can leave this out?

**Review by henningandersen — COMMENTED** [review]
Left a few comments.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        Engine engine = indexShard.getEngineOrNull();
+        if (engine == null) {
+            listener.onFailure(new AlreadyClosedException("engine closed"));
+            return;
+        }
+        var get = new Engine.Get(request.realtime(), request.realtime(), request.id()).version(request.version())
+            .versionType(request.versionType());
+        assert Objects.equals(get.uid().field(), IdFieldMapper.NAME) : get.uid().field();
+        ActionListener.completeWith(listener, () -> {
+            if (engine.isInVersionMap(get.uid().bytes())) {
```
[the remark this answers is not in the document]
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:64]: I was wondering if maybe this should just be implemented as an internal detail of `InternalEngine#get` - like a parameter on `GetResult` that is get from translog only? And it sidesteps the searcher fallback in that method?

Would make it all operate from the single call to the version map. And avoids exposing more methods on the `Engine` which is suboptimal.

The return method if it is not found in translog could be `GetResult.NOT_EXISTS` or it could be expanded a bit to clarify that only the translog was checked.

@henningandersen thoughts?

[the remark this answers is not in the document]
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:64]: Btw - this comment in `InternalEngine` is out of date right? We definitely read non-updates from translog these days I think.

```
if (get.isReadFromTranslog()) {
                        // this is only used for updates - API _GET calls will always read form a reader for consistency
                        // the update call doesn't need the consistency since it's source only + _parent but parent can go away in 7.0
```

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 9]
```diff
@@ -1982,4 +1987,9 @@ public RefreshResult(boolean refreshed) {
             this(refreshed, UNKNOWN_GENERATION);
         }
     }
+
+    // Checks whether the given doc id (provided as uid) exists in the versionMap
+    public boolean isInVersionMap(BytesRef uid) {
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java:1992]: As I mentioned in my other comment, this is okay if necessary. But I don't love adding more methods to the engine if this can be avoided.

**Review by Tim-Brooks — COMMENTED** [review]
All of Henning's comments seem sensible to me. But I had one major design question that I was curious about.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:64]: Yes. I was thinking the same after Henning's comment. Returning early from `get` is a way to also address what Henning mentioned above since we'll do everything in one go. I'm ok with doing it here or a follow up.

[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:64]: And yes, that comment is outdated. `get.isReadFromTranslog` gets the value of `get.realtime` except for one place I think (`TermVectorsService`).

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 10]
```diff
@@ -104,16 +106,69 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
...
+    private void handleGetOnUnpromotableShard(GetRequest request, IndexShard indexShard, ActionListener<GetResponse> listener)
+        throws IOException {
+        ShardId shardId = indexShard.shardId();
+        DiscoveryNode node = getCurrentNodeOfPrimary(shardId);
+        if (request.refresh()) {
+            logger.trace("send refresh action for shard {} to node {}", shardId, node.getId());
+            var refreshRequest = new BasicReplicationRequest(shardId);
+            refreshRequest.setParentTask(request.getParentTask());
+            transportService.sendRequest(
+                node,
```
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java]: done in https://github.com/elastic/elasticsearch/pull/93976/commits/101a09f601bc57269d28f0276636cff22a73b02d

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 11]
```diff
@@ -104,16 +106,69 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
...
+                TransportShardRefreshAction.NAME,
+                refreshRequest,
+                new ActionListenerResponseHandler<>(
+                    listener.map(t -> shardOperation(request, shardId)),
+                    ReplicationResponse::new,
+                    ThreadPool.Names.GET
+                )
+            );
+        } else if (request.realtime()) {
+            transportService.sendRequest(
```
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:154]: What kind of retries do you mean? Just trying the same shard again? What would be a similar example? I've tried to make it similar to `GetAction` itself, which only retries if there are other shard copies available (if I understand `TransportSingleShardAction` correctly). However, In this case, we'll have only one indexing shard.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 12]
```diff
@@ -104,16 +106,69 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
...
+                        logger.trace("received result for real-time get for id '{}' from promotable shard", request.id());
+                        l.onResponse(new GetResponse(r.getResult()));
+                    } else {
+                        assert r.segmentGeneration() > -1L;
+                        logger.trace(
+                            "no result for real-time get for id '{}' from promotable shard (segment generation to wait for: {})",
+                            request.id(),
+                            r.segmentGeneration()
+                        );
+                        indexShard.waitForSegmentGeneration(r.segmentGeneration(), l.map(aLong -> shardOperation(request, shardId)));
```
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java]: done in https://github.com/elastic/elasticsearch/pull/93976/commits/101a09f601bc57269d28f0276636cff22a73b02d

File: server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java (hunk 7 above)
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java]: done in https://github.com/elastic/elasticsearch/pull/93976/commits/101a09f601bc57269d28f0276636cff22a73b02d

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, GetRequest::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetRequest request, ActionListener<Response> listener) {
+        final ShardId shardId = request.getInternalShardId();
```
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java]: I've added a new Request type that is just a wrapper basically. But since `GetRequest` itself doesn't implement any equality/etc, I haven't done it for this either. Not sure if it worths all the boilerplate code that has to be added in several places.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:69]: I like Tim's proposal. Whether it is done here or in a follow-up is fine by me.

Yes, the comment is out of data, I've stumbled upon that a few times myself too. Updating it in this PR makes sense to me.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:154]: I'll create a follow up PR/Ticket for that.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+                        getRequest.storedFields(),
+                        getRequest.realtime(),
+                        getRequest.version(),
+                        getRequest.versionType(),
+                        getRequest.fetchSourceContext(),
+                        getRequest.isForceSyntheticSource()
+                    );
+                return new Response(result, -1);
+            }
+            long segmentGeneration = engine.commitStats().getGeneration();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:85]: I think there is a problem here - and perhaps also in the `RefreshResult`'s generation (will need to dig a bit into that).

If we cannot get rid of all regular refresh'es, a normal refresh will drop the now refreshed state from the version map, but will not advance the generation (which only advances on lucene commit/elasticsearch flush).

I see two ways out:

1. Convert all refresh'es to flush in stateless. This probably requires us to disable scheduled refresh (if not already).
2. Track that we cleared the version map due to a refresh and not due to a flush. Then trigger a flush here to remedy it. We could add a engine.getRefreshGeneration() method that returns commit generation + 1 if the version map has had a refresh.

I'm inclined to go with 2.

File: server/src/main/java/org/elasticsearch/action/support/single/shard/SingleShardRequest.java [hunk 15]
```diff
@@ -32,7 +32,7 @@
      */
     @Nullable
     protected String index;
-    ShardId internalShardId;
+    protected ShardId internalShardId;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/support/single/shard/SingleShardRequest.java:35]: I think this change is unnecessary?

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:72]: I've pushed https://github.com/elastic/elasticsearch/pull/93976/commits/3c6f60ac1a439e94c517cd1ef91038d1227b0b47

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+            return new Response(null, segmentGeneration);
+        });
+    }
+
+    public static class Request extends ActionRequest {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:90]: Let us also add a serialization test for this.

**Review by henningandersen — COMMENTED** [review]
A few more comments.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java (hunk 16 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:90]: Thanks for pointing out that `GetRequest` prevents this. For now, let us omit this.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java (hunk 14 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:85]: We discussed this some more and the fact that a flush has an embedded refresh makes option 2 difficult. We will instead track the ids affected since the last flush. We may need to only do this when tracking translog location and when we are on a promotable non-search shard. Also, reversing whether we track translog location could be beneficial in order to keep memory use at bay.

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java:85]: For now I'll park this PR. This needs to be continued once we have the mechanism to keep track of the evacuated IDs due to a refresh, in LiveVersionMap. I'll create a follow up Jira issue for that.

File: server/src/main/java/org/elasticsearch/action/support/single/shard/SingleShardRequest.java (hunk 15 above)
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/support/single/shard/SingleShardRequest.java]: Yes. I've reverted it.

**Comment by pxsalehi** [comment]
@elasticmachine test this please

hmm... seems like an infra issue!

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:99]: I think that is different. The abstract class do have subclasses that return null to have the action executed on the local node. But in this case we use `OperationRouting.getShards`, which returns a non-null instance.

I think we should remove these 3 lines (unless you had a test failure and I missed something when looking at this).

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 17]
```diff
@@ -91,6 +110,16 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
     protected void asyncShardOperation(GetRequest request, ShardId shardId, ActionListener<GetResponse> listener) throws IOException {
         IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
         IndexShard indexShard = indexService.getShard(shardId.id());
+        if (indexShard.routingEntry() == null) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:113]: Can this happen? Looks like we init it in `IndexShard` constructor and then only update it to non-null values?

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 18]
```diff
@@ -148,6 +177,67 @@ private void asyncGet(GetRequest request, ShardId shardId, ActionListener<GetRes
...
+            super.asyncShardOperation(request, shardId, listener);
+        }
+    }
+
+    private DiscoveryNode getCurrentNodeOfPrimary(ShardId shardId) {
+        var shardRoutingTable = clusterService.state().routingTable().shardRoutingTable(shardId);
+        if (shardRoutingTable.primaryShard() == null || shardRoutingTable.primaryShard().active() == false) {
+            throw new NoShardAvailableActionException(shardId, "primary shard is not active");
+        }
+        DiscoveryNode node = clusterService.state().nodes().get(shardRoutingTable.primaryShard().currentNodeId());
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:234]: Can we use the same `ClusterState` instance here rather than asking for it again? That way, the null check below can turn into an assertion (`assert node != null`).

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java (hunk 17 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:113]: Yeah, but that looks like both comes from `TransportIndicesStatsAction`, which was done in 2014 and thus on a very different code base. I think we should assume this does not happen, since we assume so in other actions, for instance `TransportGetFromTranslogAction`, `PostWriteRefresh` and many others. Unless you know where it can happen I suggest to remove the new null check, we do not want null-checks all over for things that cannot be null.

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java]: I agree!

File: server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:98]: To me it seems `preferenceActiveShardIterator` which is used by `OperationRouting.getShards` can return null. `useOnlyPromotableShardsForStateless` which also uses its return value, has some null check as well.

[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:98]: Would that not also have NPE'd today? We would then execute it "locally", which may not have the shard and thus NPE.

But let us keep it for now. Perhaps add a comment on `preferenceActiveShardIterator` to clarify the case.

[the remark this answers is not in the document]
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:98]: > Would that not also have NPE'd today? 

I think not because we are not dereferencing it. We just return it and `AsyncSingleAction` takes care of it. In this PR, we use it to filter the shards.
